File: site/mp2i/files/C10/r90.py
DEAD = '.'
ALIVE = '#'

def evolution(config):
    n = len(config)
    nconfig = config[0]
    for i in range(1,n-1):
        if (config[i+1]==config[i-1]):
            nconfig += DEAD
        else:
            nconfig += ALIVE
    nconfig += config[-1]
    return nconfig

n = 30
c = "."*(n//2)+"#"+"."*(n//2-1)
s = 0
seen = set()
seen.add(c)
step = 0
end = False
while (not end):
    step += 1
    c = evolution(c)
    if c not in seen:
        seen.add(c)
    else:
        end = True
print(c)
print(step)

File: mp2i/files/C10/r90.c
#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>
#include <string.h>

#define DEAD '.'
#define ALIVE '#'

char evo_cell(char prev,  char next)
{
    if (prev != next)
    {
       return ALIVE; 
    }
    return DEAD;
}

void evo_config(char *config)
{
    int n = strlen(config);
    char* temp = malloc(sizeof(char)*n);
    temp[0] = config[0];
    temp[n-1] = config[n-1];
    for (int i=1;i <n-1;i++)
    {
        temp[i] = evo_cell(config[i-1],config[i+1]);
    }
    for (int i=0;i <n;i++)
    {
        config[i] = temp[i];
    }
    free(temp);
}

bool same(char *c, char* r, int n)
{
    for (int i=0;i<n;i++)
    {
        if (c[i]!=r[i])
        {
            return false;
        }
    }
    return true;
}

int main()
{
    int n= 30;
    char* c = malloc(sizeof(char)*n);
    char* r = malloc(sizeof(char)*n);
    for (int i=0;i<n;i++)
    {
        c[i] = DEAD;
        r[i] = DEAD;
    }
    c[n/2] = ALIVE;
    r[n/2] = ALIVE;
    int step = 0;
    bool end = false;
    while (!end)
    {
        evo_config(c);
        if (same(r,c,n))
        {
            end = true;
        }
        step += 1;
    }
    printf("Cycle après %d étapes \n",step);
    free(c);
}